File: celesta-test/score/testTable/testTable.py
# coding=UTF-8

from _testTable_orm import tBlobCursor, tXRecCursor, tCsvLineCursor, tIterateCursor, \
    tCopyFieldsCursor, tLimitCursor, tWithDecimalCursor, tWithDateTimeZCursor

from ru.curs.celesta import CelestaException
from ru.curs.celesta.unit import TestClass, CelestaTestCase

from java.sql import Timestamp
from java.time import LocalDateTime, Month, ZonedDateTime, ZoneId
from java.math import BigDecimal
from java.util import TimeZone

import java.io.OutputStreamWriter as OutputStreamWriter
import java.io.InputStreamReader as InputStreamReader
import java.io.BufferedReader as BufferedReader

@TestClass
class TestTable(CelestaTestCase):

    def test_calc_blob(self):
        cursor = tBlobCursor(self.context)
        cursor.deleteAll()

        cursor.insert()
        cursor.get(1)

        self.assertEquals(1, cursor.id)
        self.assertEquals(None, cursor.dat)

        cursor.calcdat()

        self.assertTrue(cursor.dat and cursor.dat.isNull())

        os = cursor.dat.getOutStream()
        osw = OutputStreamWriter(os, 'utf-8')
        try:
            osw.append('blob field')
        finally:
            osw.close()

        cursor.update()
        cursor.clear()
        cursor.get(1)
        cursor.calcdat()
        bf = BufferedReader(InputStreamReader(cursor.dat.getInStream(), 'utf-8'))
        self.assertEquals('blob field', bf.readLine())
        bf.close()

        cursor.clear()
        cursor.calcdat()
        os = cursor.dat.getOutStream()
        osw = OutputStreamWriter(os, 'utf-8')
        try:
            osw.append('blob field 2!')
        finally:
            osw.close()
        cursor.insert()

        cursor.clear()
        cursor.get(2)
        cursor.calcdat()
        bf = BufferedReader(InputStreamReader(cursor.dat.getInStream(), 'utf-8'))
        self.assertEquals('blob field 2!', bf.readLine())
        bf.close()

    def test_getXRec(self):
        cursor = tXRecCursor(self.context)
        cursor.deleteAll()

        id = 1
        num = 10
        cost = 10.2
        title = 'product'
        isActive = True
        created = Timestamp.valueOf(LocalDateTime.of(2018, Month.of(1), 11, 19, 15))

        cursor.num = num
        cursor.cost = cost
        cursor.title = title
        cursor.isActive = isActive
        cursor.created = created

        xRec = cursor.getXRec()
        self._assertXRecCursorFields(xRec, None, None, None, None, None, None)
        cursor.insert()

        self._assertXRecCursorFields(xRec, id, num, cost, title, isActive, created)
        cursor.clear()

        xRec = cursor.getXRec()
        self._assertXRecCursorFields(xRec, None, None, None, None, None, None)

        cursor.get(1)
        self._assertXRecCursorFields(xRec, id, num, cost, title, isActive, created)

        cursor.num = num + 1
        cursor.cost = cost + 1
        cursor.title = title + 'asd'
        cursor.isActive = False
        cursor.created = Timestamp.valueOf(LocalDateTime.of(2017, Month.of(1), 11, 19, 15))

        self._assertXRecCursorFields(xRec, id, num, cost, title, isActive, created)

    def test_asCSVLine(self):
        cursor = tCsvLineCursor(self.context)
        self.assertEquals('NULL,NULL',cursor.asCSVLine())

        cursor.id = 1
        self.assertEquals('1,NULL',cursor.asCSVLine())

        cursor.title = 'noQuotes'
        self.assertEquals('1,noQuotes',cursor.asCSVLine())

        cursor.title = '"withQuotes"'
        self.assertEquals('1,"""withQuotes"""',cursor.asCSVLine())

        cursor.title = None
        self.assertEquals('1,NULL',cursor.asCSVLine())

    def test_iterate(self):
        cursor = tIterateCursor(self.context)
        cursor.insert()
        cursor.clear()
        cursor.insert()

        idList = []

        for cursor in cursor.iterate():
            idList.append(cursor.id)

        self.assertEquals(2, idList.__len__())
        self.assertEquals(1, idList[0])
        self.assertEquals(2, idList[1])

    def test_CopyFieldsFrom(self):
        cursor = tCopyFieldsCursor(self.context)
        cursorFrom = tCopyFieldsCursor(self.context)

        id = 11234
        title = 'ttt'

        cursorFrom.id = id
        cursorFrom.title = title
        cursor.copyFieldsFrom(cursorFrom)

        self.assertEquals(id, cursor.id)
        self.assertEquals(title, cursor.title)

    def test_limit(self):
        cursor = tLimitCursor(self.context)

        for i in range(3):
            cursor.insert()
            cursor.clear()

        idList = []

        cursor.limit(0, 2)
        for cursor in cursor.iterate():
            idList.append(cursor.id)
        self.assertEquals(2, idList.__len__())
        self.assertEquals(1, idList[0])
        self.assertEquals(2, idList[1])
        del idList[:]

        cursor.limit(2, 1)
        for cursor in cursor.iterate():
            idList.append(cursor.id)
        self.assertEquals(1, idList.__len__())
        self.assertEquals(3, idList[0])
        del idList[:]

        cursor.limit(3, 0)
        for cursor in cursor.iterate():
            idList.append(cursor.id)
        self.assertEquals(0, idList.__len__())
        del idList[:]

        cursor.limit(3, 5)
        for cursor in cursor.iterate():
            idList.append(cursor.id)
        self.assertEquals(0, idList.__len__())

    def test_decimal(self):
        c = tWithDecimalCursor(self.context)

        c.insert()
        c.first()
        self.assertEquals(BigDecimal('5.2'), c.cost.stripTrailingZeros())

        c.cost = BigDecimal('5.289')
        c.update()
        c.first()
        self.assertEquals(BigDecimal('5.29'), c.cost.stripTrailingZeros())

        c.cost = BigDecimal('123.2')
        update = lambda : c.update()
        self.assertThrows(CelestaException, update)

    def test_datetime_with_time_zone(self):
        oldDefaultTimeZone = TimeZone.getDefault()
        try:
            TimeZone.setDefault(TimeZone.getTimeZone("GMT+4"))
            c = tWithDateTimeZCursor(self.context)

            zoneId = ZoneId.of("GMT+2")
            #This is the datetime we will insert
            localDateTime = LocalDateTime.of(2017, Month.DECEMBER, 31, 22, 0, 0)
            zonedDateTime = ZonedDateTime.of(localDateTime, zoneId)
            #This is the datetime we expect the database to receive
            utcDateTime = LocalDateTime.of(2018, Month.JANUARY, 1, 0, 0, 0)

            c.eventDate = zonedDateTime
            c.insert()
            c.clear()

            c.first()

            zoneIdAfterSelect = ZoneId.of("GMT+4")
            self.assertEquals(utcDateTime, c.eventDate.toLocalDateTime())
            self.assertEquals(zoneIdAfterSelect, c.eventDate.getZone())

        finally:
            TimeZone.setDefault(oldDefaultTimeZone)


    def _assertXRecCursorFields(self, cursor, id, num, cost, title, isActive, created):
        self.assertEquals(id, cursor.id)
        self.assertEquals(num, cursor.num)
        self.assertEquals(cost, cursor.cost)
        self.assertEquals(title, cursor.title)
        self.assertEquals(isActive, cursor.isActive)
        self.assertEquals(created, cursor.created)

File: celesta-test/src/test/java/ru/curs/celesta/script/TestTable.java
package ru.curs.celesta.script;


import org.junit.jupiter.api.TestTemplate;
import ru.curs.celesta.CallContext;
import ru.curs.celesta.CelestaException;
import ru.curs.celesta.DBType;
import testTable.*;

import java.io.*;
import java.math.BigDecimal;
import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.time.Month;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.TimeZone;

import static junit.framework.TestCase.assertNotNull;
import static org.junit.jupiter.api.Assertions.*;

public class TestTable implements ScriptTest {

    @TestTemplate
    public void test_calc_blob(CallContext cc) throws IOException {

        TBlobCursor cursor = new TBlobCursor(cc);
        cursor.deleteAll();

        cursor.insert();
        cursor.get(1);

        assertEquals(1, cursor.getId().intValue());
        assertNull(cursor.getDat());

        cursor.calcDat();

        assertNotNull(cursor.getDat());
        assertTrue(cursor.getDat().isNull());

        OutputStream os = cursor.getDat().getOutStream();
        try (OutputStreamWriter osw = new OutputStreamWriter(os, "utf-8")) {
            osw.append("blob field");
        }

        cursor.update();
        cursor.clear();
        cursor.get(1);
        cursor.calcDat();
        try (BufferedReader bf = new BufferedReader(
                new InputStreamReader(cursor.getDat().getInStream(), "utf-8"))) {
            assertEquals("blob field", bf.readLine());
        }

        cursor.clear();
        cursor.calcDat();
        os = cursor.getDat().getOutStream();
        try (OutputStreamWriter osw = new OutputStreamWriter(os, "utf-8")) {
            osw.append("blob field 2!");
        }

        cursor.insert();

        cursor.clear();
        cursor.get(2);
        cursor.calcDat();
        try (BufferedReader bf = new BufferedReader(
                new InputStreamReader(cursor.getDat().getInStream(), "utf-8"))) {
            assertEquals("blob field 2!", bf.readLine());
        }
    }

    @TestTemplate
    public void test_getXRec(CallContext cc) {
        TXRecCursor cursor = new TXRecCursor(cc);
        cursor.deleteAll();

        int id = 1;
        int num = 10;
        double cost = 10.2;
        String title = "product";
        boolean isActive = true;
        Timestamp created = Timestamp.valueOf(LocalDateTime.of(2018, Month.of(1), 11, 19, 15));

        cursor.setNum(num);
        cursor.setCost(cost);
        cursor.setTitle(title);
        cursor.setIsActive(isActive);
        cursor.setCreated(created);

        //TODO: type cast!
        TXRecCursor xRec = (TXRecCursor) cursor.getXRec();
        assertXRecCursorFields(xRec, null, null, null, null, null, null);
        cursor.insert();

        //TODO: WTF? worked in Python, but I cannot see why.
        //@ioanngolovko maybe just ignore this
        assertXRecCursorFields(xRec, id, num, cost, title, isActive, created);
        cursor.clear();

        xRec = (TXRecCursor) cursor.getXRec();
        assertXRecCursorFields(xRec, null, null, null, null, null, null);

        cursor.get(1);
        assertXRecCursorFields(xRec, id, num, cost, title, isActive, created);

        cursor.setNum(num + 1);
        cursor.setCost(cost + 1.0);
        cursor.setTitle(title + "asd");
        cursor.setIsActive(false);
        cursor.setCreated(Timestamp.valueOf(LocalDateTime.of(2017, Month.of(1), 11, 19, 15)));

        assertXRecCursorFields(xRec, id, num, cost, title, isActive, created);
    }

    @TestTemplate
    public void test_asCSVLine(CallContext cc) {
        TCsvLineCursor cursor = new TCsvLineCursor(cc);
        assertEquals("NULL,NULL", cursor.asCSVLine());

        cursor.setId(1);
        assertEquals("1,NULL", cursor.asCSVLine());

        cursor.setTitle("noQuotes");
        assertEquals("1,noQuotes", cursor.asCSVLine());

        cursor.setTitle("\"withQuotes\"");
        assertEquals("1,\"\"\"withQuotes\"\"\"", cursor.asCSVLine());

        cursor.setTitle(null);
        assertEquals("1,NULL", cursor.asCSVLine());
    }

    @TestTemplate
    public void test_iterate(CallContext cc) {
        TIterateCursor cursor = new TIterateCursor(cc);
        cursor.insert();
        cursor.clear();
        cursor.insert();

        ArrayList<Integer> idList = new ArrayList<>();
        for (TIterateCursor c : cursor) {
            idList.add(c.getId());
        }

        assertEquals(2, idList.size());
        assertEquals(1, idList.get(0).intValue());
        assertEquals(2, idList.get(1).intValue());
    }

    @TestTemplate
    public void test_CopyFieldsFrom(CallContext cc) {
        TCopyFieldsCursor cursor = new TCopyFieldsCursor(cc);

        TCopyFieldsCursor cursorFrom = new TCopyFieldsCursor(cc);

        int id = 11234;
        String title = "ttt";

        cursorFrom.setId(id);
        cursorFrom.setTitle(title);
        cursor.copyFieldsFrom(cursorFrom);

        assertEquals(id, cursor.getId().intValue());
        assertEquals(title, cursor.getTitle());
    }

    @TestTemplate
    public void test_limit(CallContext cc) {
        TLimitCursor cursor = new TLimitCursor(cc);

        for (int i = 0; i < 3; i++) {
            cursor.insert();
            cursor.clear();
        }

        ArrayList<Integer> idList = new ArrayList<>();

        cursor.limit(0, 2);
        for (TLimitCursor c : cursor) {
            idList.add(c.getId());
        }

        assertEquals(2, idList.size());
        assertEquals(1, idList.get(0).intValue());
        assertEquals(2, idList.get(1).intValue());
        idList.clear();

        cursor.limit(2, 1);
        for (TLimitCursor c : cursor) {
            idList.add(c.getId());
        }
        assertEquals(1, idList.size());
        assertEquals(3, idList.get(0).intValue());

        idList.clear();


        cursor.limit(3, 0);
        for (TLimitCursor c : cursor) {
            idList.add(c.getId());
        }
        assertEquals(0, idList.size());
        idList.clear();

        cursor.limit(3, 5);
        for (TLimitCursor c : cursor) {
            idList.add(c.getId());
        }
        assertEquals(0, idList.size());
    }

    @TestTemplate
    public void test_decimal(CallContext cc) {
        TWithDecimalCursor c = new TWithDecimalCursor(cc);

        c.insert();
        c.first();
        assertEquals(new BigDecimal("5.2"), c.getCost().stripTrailingZeros());

        c.setCost(new BigDecimal("5.289"));
        c.update();
        c.first();
        assertEquals(new BigDecimal("5.29"), c.getCost().stripTrailingZeros());

        c.setCost(new BigDecimal("123.2"));
        //TODO: known non-conformity
        if (DBType.FIREBIRD.equals(cc.getDbAdaptor().getType())) {
            c.update();
            c.first();
            assertEquals(new BigDecimal("123.2"), c.getCost().stripTrailingZeros());
        } else{
            assertThrows(CelestaException.class, () -> c.update());
        }

        c.setCost(new BigDecimal("1234.25235"));

        if (DBType.FIREBIRD.equals(cc.getDbAdaptor().getType())) {
            c.update();
            c.first();
            assertEquals(new BigDecimal("1234.25"), c.getCost().stripTrailingZeros());
        } else{
            assertThrows(CelestaException.class, () -> c.update());
        }
    }

    @TestTemplate
    public void test_datetime_with_time_zone(CallContext cc) {
        TimeZone oldDefaultTimeZone = TimeZone.getDefault();
        try {
            TimeZone.setDefault(TimeZone.getTimeZone("GMT+4"));
            TWithDateTimeZCursor c = new TWithDateTimeZCursor(cc);

            ZoneId zoneId = ZoneId.of("GMT+2");
            /*
            This is
            the datetime
            we will
            insert*/
            LocalDateTime localDateTime = LocalDateTime.of(2017, Month.DECEMBER, 31, 22, 0, 0);
            ZonedDateTime zonedDateTime = ZonedDateTime.of(localDateTime, zoneId);
            /*
            This is
            the datetime
            we expect
            the database
            to receive*/
            LocalDateTime utcDateTime = LocalDateTime.of(2018, Month.JANUARY, 1, 0, 0, 0);

            c.setEventDate(zonedDateTime);
            c.insert();
            c.clear();

            c.first();

            ZoneId zoneIdAfterSelect = ZoneId.of("GMT+4");
            assertEquals(utcDateTime, c.getEventDate().toLocalDateTime());
            assertEquals(zoneIdAfterSelect, c.getEventDate().getZone());

        } finally {
            TimeZone.setDefault(oldDefaultTimeZone);
        }

    }

    private void assertXRecCursorFields(TXRecCursor cursor, Integer id, Integer num, Double cost, String title, Boolean isActive, Timestamp created) {
        assertEquals(id, cursor.getId());
        assertEquals(num, cursor.getNum());
        assertEquals(cost, cursor.getCost());
        assertEquals(title, cursor.getTitle());
        assertEquals(isActive, cursor.getIsActive());
        assertEquals(created, cursor.getCreated());
    }

    @TestTemplate
    public void test_setRange(CallContext cc) {

        TXRecCursor cursor = new TXRecCursor(cc);
        cursor.deleteAll();

        cursor.setNum(11);
        cursor.insert();

        cursor.clear();
        cursor.setNum(22);
        cursor.insert();

        cursor.clear();
        cursor.setNum(33);
        cursor.insert();

        cursor = new TXRecCursor(cc);
        assertEquals(3, cursor.count());

        cursor.setRange(cursor.COLUMNS.num(), 22);
        assertEquals(1, cursor.count());

        cursor.setRange(cursor.COLUMNS.num());
        assertEquals(3, cursor.count());

        cursor.setRange(cursor.COLUMNS.num(), 22, 33);
        assertEquals(2, cursor.count());
    }

    @TestTemplate
    public void test_setFilter(CallContext cc) {

        TXRecCursor cursor = new TXRecCursor(cc);
        cursor.deleteAll();

        cursor.setNum(11);
        cursor.insert();

        cursor.clear();
        cursor.setNum(22);
        cursor.insert();

        cursor.clear();
        cursor.setNum(33);
        cursor.insert();

        cursor = new TXRecCursor(cc);
        assertEquals(3, cursor.count());

        cursor.setFilter(cursor.COLUMNS.num(), "22");
        assertEquals(1, cursor.count());

        cursor.setRange(cursor.COLUMNS.num());
        assertEquals(3, cursor.count());

        cursor.setFilter(cursor.COLUMNS.num(), "22..33");
        assertEquals(2, cursor.count());
    }

}
